src/App.tsx

```diff
@@ -33,7 +33,28 @@ import {
 import { colorSchemeAtom } from './common/colors';
 import { useCurrentUser } from './common/hooks/useCurrentUser';
 import { useRefetch } from './common/hooks/useRefetch';
+import { PreventNavigationModal } from './components/PreventNavigationModal';
+import { isNavigationModalVisibleAtom } from './common/hooks/usePreventNavigation';
 
+interface PreventLeavingPage {
+  prevent: boolean;
+  actionKey?: 'switchCompany' | 'browserBack';
```

Comment:
What exactly is actionKey and what it is used for?

Reply:
@beganovich We have two cases when the `actionKey` is needed. The `actionKey` is a property that defines what is currently triggered in our case—either the switch company functionality or triggering the back button on the browser.

So, why is it important?!

The functionality of switching companies is based on the browser `reloading`, which immediately means that if needed at that moment, the `beforeunload` event will be triggered, and the browser will show the default modal for preventing leaving the page. This modal provides 'Cancel' and 'Leave' buttons, but in the case of switching companies, 'Cancel' will work just as we need. However, the 'Leave' button will not know how to switch the company (it will not remember the 'switchCompany' function and execute it as needed). Therefore, we need to display our own modal and also prevent the default one from being displayed. Essentially, the 'switchCompany' actionKey is only used when it is necessary to prevent the display of the default browser's modal.

The second 'browserBack' action key is used when it is necessary to push a new state into the history. I conducted thorough research, and it is impossible to disable the back button of the browser unless we push the current page into the state. This way, when the user clicks the back button, they will stay on that page. So, this actionKey is used when we need to do that, avoiding unnecessary pushing of a new state into the history. Additionally, there is another case when it is also very important. If the user triggers the back button and then triggers the 'continue editing' button on the prevent navigation modal multiple times in a row (potentially), they will have, let's say, 10+ instances of the current page in the history object. Therefore, when the prevention of navigation is no longer needed, the user would have to click the back button 10+ times in the browser to go back to the previously non-prevented page. To avoid this poor UX, I store the last page once the user is not prevented from navigation, and then I navigate the user to that page instead of navigating to the exact same page where they are at that moment.

Please let me know if you need any additional explanation for this part of the logic.